src/storage/transaction/TossEdgeIterator.h

```diff
@@ -0,0 +1,220 @@
+                                                                   iter_->val().str(),
+                                                                   recoverEdges_.back()));
+                lastRank_ = NebulaKeyUtils::getRank(planContext_->vIdLen_, rawKey);
+                lastDstId_ = NebulaKeyUtils::getDstId(planContext_->vIdLen_, rawKey).str();
+                lastIsLock_ = true;
+                continue;
+            } else {
+                LOG_IF(INFO, FLAGS_trace_toss)
+                    << "next to a weird record: \n"
+                    << folly::hexDump(iter_->key().data(), iter_->key().size())
+                    << folly::hexDump(iter_->val().data(), iter_->val().size());
+            }
+        }
+        LOG_IF(INFO, FLAGS_trace_toss) << "no more iter to read";
+
+        if (needWaitResumeTask_) {
+            folly::collectAll(resumeTasks_).wait();
+            needWaitResumeTask_ = false;
+            recoverEdgesIter_ = recoverEdges_.begin();
+            firstLoop_ = true;
+        }
+
+        while (recoverEdgesIter_ != recoverEdges_.end()) {
+            if (recoverEdgesIter_ != recoverEdges_.begin() || firstLoop_ != true) {
+                recoverEdgesIter_++;
+            }
+            if (recoverEdgesIter_ == recoverEdges_.end()) {
+                return;
+            }
+            firstLoop_ = false;
+            auto data = (*recoverEdgesIter_)->copy();
+            if (!data.second.empty()) {
+                setReader(data.second);
+            }
+        }
+    }
+
+    bool isEdge(const folly::StringPiece& key) {
+        return NebulaKeyUtils::isEdge(planContext_->vIdLen_, key);
+    }
+
+    bool isLatestEdge(const folly::StringPiece& key) {
+        auto rank = NebulaKeyUtils::getRank(planContext_->vIdLen_, key);
+        auto dstId = NebulaKeyUtils::getDstId(planContext_->vIdLen_, key);
+        return !(rank == lastRank_ && dstId == lastDstId_);
+    }
+
+    bool setReader(folly::StringPiece val) {
+        reader_.reset();
+        if (!reader_) {
+            reader_.reset(*schemas_, val);
+            if (!reader_) {
+                planContext_->resultStat_ = ResultStatus::ILLEGAL_DATA;
+                return false;
+            }
+        } else if (!reader_->reset(*schemas_, val)) {
+            planContext_->resultStat_ = ResultStatus::ILLEGAL_DATA;
+            return false;
+        }
+
+        if (hasTtl_ && CommonUtils::checkDataExpiredForTTL(
+                                            schemas_->back().get(), reader_.get(),
+                                            ttlCol_, ttlDuration_)) {
+            reader_.reset();
+            return false;
+        }
+        return true;
+    }
+
+private:
+    // using shared_ptr because this Iterator may be deleted if there is a limit in nGQL
+    // using folly::Synchronized because the scan thread and the resume thread may write same ptr
+    using TResultsItem = std::shared_ptr<folly::Synchronized<KV>>;
```

Comment:
No need to use `Synchronized` for now?

Reply:
no, let's say there is a lock and an edge. and thread1 is scanning.
when meet the lock, thread1 let thread2 do the lock resuming, and continue.
then it meet that edge. and try to write the props, 
but this time, thread2 may return, and want to write to the same locating.

Reply:
You're right. That's why you check if TResultsItem is empty, because in worst case, thread 2 may return at first before thread 1 write it.